File: solana/src/client/WorkflowStep.ts
import { field, vec } from '@dao-xyz/borsh'
import BN from 'bn.js'
import { BorshPubkey } from './BorshPubkey'

export class WorkflowStep {
  constructor(init: WorkflowStep) {
    // Object.assign(this, init)
    this.stepId = init.stepId
    // this.fromToken = init.fromToken
    this.amount = init.amount
    this.amountIsPercent = init.amountIsPercent
    this.stepArgs = init.stepArgs
  }

  @field({ type: 'u16' })
  stepId: number
  // @field(BorshPubkey)
  // fromToken: string
  @field({ type: 'u64' })
  amount: BN
  @field({ type: 'bool' })
  amountIsPercent: boolean
  @field({ type: vec('u8') })
  stepArgs: Uint8Array
}

File: solana/src/program/src/engine/workflow_step.rs
use borsh::maybestd::vec::Vec;
use borsh::BorshDeserialize;
use solana_program::account_info::AccountInfo;
use solana_program::pubkey::Pubkey;

#[derive(BorshDeserialize)]
pub struct WorkflowStepArgs {
    pub step_id: u16,
    // pub from_token: Pubkey,
    pub amount: u64,
    pub amount_is_percent: bool,
    pub step_args: Vec<u8>,
}

#[derive(BorshDeserialize)]
pub struct WorkflowArgs {
    pub steps: borsh::maybestd::vec::Vec<WorkflowStepArgs>,
}

pub struct WorkflowStepResult {
    pub from_mint_address: Option<Pubkey>,
    pub to_mint_address: Option<Pubkey>,
    pub amount: u64,
}

pub trait WorkflowStep {
    // return the 'from' account for this step
    fn name(&self) -> &'static str;
    fn deserialize_step_data(&mut self, step_args: &Vec<u8>);
    fn get_from_mint_address(&self, accounts: &[AccountInfo]) -> Pubkey;
    fn execute(&self, program_id: &Pubkey, accounts: &[AccountInfo], amount: u64) -> WorkflowStepResult;
}
